src/refresh/RefreshTimer.ts

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+const MinimumBackoffInMs = 30 * 1000; // 30s
+const MaximumBackoffInMs = 10 * 60 * 1000; // 10min
+const MaxAttempts = 63;
+const JitterRatio = 0.25;
+
+export class RefreshTimer {
+    private _minBackoff: number;
+    private _maxBackoff: number;
+    private _attempts: number;
+    private _backoffEnd: number; // Timestamp
+    constructor(
+        private _interval: number
+    ) {
+        this._minBackoff = Math.min(this._interval, MinimumBackoffInMs);
+        this._maxBackoff = Math.min(this._interval, MaximumBackoffInMs);
+        this._attempts = 0;
+        this._backoffEnd = Date.now() + this._interval;
+    }
+
+    public isDuringBackoff(): boolean {
+        return Date.now() < this._backoffEnd;
+    }
+
+    public backoff(): void {
+        this._backoffEnd = Date.now() + this._calculateBackoffTime();
+        this._attempts += 1;
+    }
+
+    public reset(): void {
+        this._backoffEnd = Date.now() + this._interval;
+        this._attempts = 0;
+    }
+
+    private _calculateBackoffTime(): number {
+        let minBackoffMs: number;
+        let maxBackoffMs: number;
+        if (this._interval <= this._minBackoff) {
+            return this._interval;
+        }
+
+        // _minBackoff <= _interval
+        if (this._interval <= this._maxBackoff) {
+            minBackoffMs = MinimumBackoffInMs
+            maxBackoffMs = this._interval;
+        } else {
+            minBackoffMs = MinimumBackoffInMs;
+            maxBackoffMs = MaximumBackoffInMs;
+        }
+
+        // exponential
+        let calculatedBackoffMs = Math.max(1, minBackoffMs) * (1 << Math.min(this._attempts, MaxAttempts));
+        if (calculatedBackoffMs > maxBackoffMs) {

```

Comment:
Is there any risk of overflow for `calculatedBackoffMs`?

Reply:
The overflow is not as critical as in other language.  Numbers (double) larger than [Number.MAX_VALUE](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_VALUE) is `infinity` which is larger than any number in comparison.

And [Number.MAX_SAFE_INTEGER](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER) is the largest **integer** that ensures comparison to work correctly.

for simplicity, current logic to check `calculatedBackoffMs > maxBackoffMs` should work even when overflow occurs, AFAIK. 

But yes, it's better to add overflow guardian, using MAX_SAFE_INTEGER. I'll update accordingly.

Reply:
Just revisit the code, if "overflow" occurs and calculatedBackoffMs  > MAX_SAFE_INTEGER, it is certainly larger than maxBackoffMs.
So, no change is needed.